File: cpp/424_longest_repeating_character_replacement.cpp
#include <array>
#include <string>

using namespace std;

class Solution {
 public:
  int characterReplacement(string s, int k) {
    array<int, 26> letter_freq{};
    int max_freq = 0;
    int start = 0, max_len = 0;
    for (int end = 0; end < s.size(); end++) {
      max_freq = max(max_freq, ++letter_freq[s[end] - 'A']);
      if (end - start + 1 - max_freq > k) {
        letter_freq[s[start] - 'A']--;
        start++;
      }
      max_len = max(max_len, end - start + 1);
    }

    return max_len;
  }
};

File: rust/424_longest_repeating_character_replacement.rs
use std::cmp::max;

impl Solution {
    pub fn character_replacement(s: String, k: i32) -> i32 {
        let mut letter_freq = [0; 26];
        let mut max_freq = 0;
        let (mut start, mut max_len) = (0, 0);
        let s_bytes = s.as_bytes();

        for (end, &c) in s_bytes.iter().enumerate() {
            let idx = (c - b'A') as usize;
            letter_freq[idx] += 1;
            max_freq = max(max_freq, letter_freq[idx]);
            if end - start + 1 - max_freq > k as usize {
                letter_freq[(s_bytes[start] - b'A') as usize] -= 1;
                start += 1;
            }
            max_len = max(max_len, end - start + 1);
        }

        max_len as i32
    }
}
